### ISSUE TITLE
switch input configuration bug

### ISSUE THREAD

**Issue by sunbowch** [braingram/dSPIN#1]
Hi!

The setSwitchMode & getSwitchMode functions in dSPINConfig.cpp have the wrong mask: 0x0100 must be 0x0010... As explained in the comment...

**Comment by braingram** [comment]
Thanks for the catch! It should be fixed now. Let me know if you see anything else (or if this fix doesn't work).

**Comment by sunbowch** [comment]
Hi!

Obviously it works for me... I'm building an animated dress exhibition 
with 12 motors and a belt reduction, so  I'm in need of the switch input 
to get the position of the dress...

Thanks for the library, by daisy-chaining the motors I can have a 
perfect sync...

**Comment by braingram** [comment]
Nice! Mind sending a link? Are you using an off-the-shelf dSpin board or did you roll your own? I've had lots of luck with [these](http://www.digikey.com/product-detail/en/stmicroelectronics/X-NUCLEO-IHM02A1/497-16251-ND/5768392) but it only daisy chains 2x per shield (with 4 chip selects giving max 8 motors).

**Comment by sunbowch** [comment]
I've used a clone of this <https://www.sparkfun.com/products/11611>, 
only a lot cheaper. The only dual controllers that I have sourced before 
hadn't the switch pin wired... I think that you can also daisy chain 
these dual cards to have more than 2 controllers on the chain, it's only 
a wiring question... So as long as you don't stack them

The project will be built next week... Customer request is only to 
rotate the dress, but I'll try to convince the customer that we can 
build a real show... My idea is to use a raspberry pi with QLC+, and 
build a driver to control the motors with the DMX values that I can grab 
on the RPi.

Then the creative part can start as I can control sound, light an motor 
control with QLC+, which is a quite good and open source software to 
create ligthshows.

The arduino in the project isn't really necessary, but it's cheaper than 
porting the libraries, and it unloads the RPi which is just at the limit 
with QLC+ if you don't disable either the display or the web server... 
One great feature of QLC+ is that you can control it through this server...

I'll send you a link as soon as it's running... And I have some pictures.

**Comment by sunbowch** [comment]
Hi!

Here is some pictures <https://goo.gl/photos/ngps1hCHnXFH8s6p7> of the 
result. All the chips are daisy-chained.

Lots of wires as there is 4 for each motor + 3 for the position sensor 
(I've used a Hall effect sensor).

Regarding the library code I had to ajust the NDSPINS value into the 
library, despite defining it in my code before the call to the 
library... Haven't found why...

I have also added some functions to set and get the BEMF compensation 
curve. (ST_SLP, INT_SPD, FN_SLP_ACC&DEC).

And 1 pin to grab the Flag pin and check for errors when it is activated.

The support functions for INT_SPD (INT_SPEED in the datasheet) have the 
wrong factor: It should be 16.777216 according to the datasheet.

The datasheet is unclear about the values of the slopes for the above 
parameters: it should be a byte integer which gives the slope value when 
multiplied by the 0.000015 resolution.

The software sends the commands to all chips, then wait until the busy 
pin desactivate or until the duration of the effect is over (typically 
for dspin_run command)... Then it either sends a softstop, or the next 
command, depending if there is a stop time after the effect. So it is 
possible to change the speed dynamically without stopping the system or 
having it to stop in a given position after some continuous turns.

I'm using the 1/128 microstepping rate to have a fluid movement at low 
speed and less noise. The most annoying problem was that the ABS_POS 
value is rapidly overflowed with such a high microstepping rate... It 
takes a few seconds at maximum speed in my case. So I have a function 
that activates every couple of seconds and after each move to detect if 
the ABS_POS register has changed sign and in this case the "0" reference 
position is recalculated and shifted the right number of steps in order 
that the position of my mannequin stays always the same ( 0° is always 
facing the same direction).

All the instructions are grouped in one int array with 3 parameters for 
each cue: command type, speed (negative is for the other direction), and 
position in degree) for every motor. Another array defines the timings 
(duration of the cue and pause after). These arrays could (and should) 
be in program memory.
Depending on the command the degree position is either relative to the 
present position or absolute.

I will add a video if I take time to make one...

The code is at your disposal... But needs some more comments and cleaning.

Sincerely,

Luc
